aws/iam_service_linked_role.go

```diff
@@ -0,0 +1,191 @@
+	})
+	if err != nil {
+		return gruntworkerrors.WithStackTrace(err)
+	}
+
+	// Wait for the deletion to complete
+	time.Sleep(3 * time.Second)
+
+	var deletionStatus *iam.GetServiceLinkedRoleDeletionStatusOutput
+
+	done := false
+	for !done {
+		done = true
+		// Check if the deletion is complete
+		deletionStatus, err = svc.GetServiceLinkedRoleDeletionStatus(&iam.GetServiceLinkedRoleDeletionStatusInput{
+			DeletionTaskId: deletionData.DeletionTaskId,
+		})
+		if err != nil {
+			return gruntworkerrors.WithStackTrace(err)
+		}
+		if aws.StringValue(deletionStatus.Status) == "IN_PROGRESS" {
+			logging.Logger.Debugf("Deletion of IAM ServiceLinked Role %s is still in progress", aws.StringValue(roleName))
+			done = false
+			time.Sleep(3 * time.Second)
+		}
+
+	}
+
+	if aws.StringValue(deletionStatus.Status) != "SUCCEEDED" {
+		err := fmt.Sprintf("Deletion of IAM ServiceLinked Role %s failed with status %s", aws.StringValue(roleName), aws.StringValue(deletionStatus.Status))
+		return gruntworkerrors.WithStackTrace(errors.New(err))
+	}
+
+	logging.Logger.Infof("Successfully Deleted IAM ServiceLinked Role %s", aws.StringValue(roleName))
+	return nil
+}
+
+// Delete all IAM Roles
+func nukeAllIamServiceLinkedRoles(session *session.Session, roleNames []*string) error {
+	region := aws.StringValue(session.Config.Region)
+	svc := iam.New(session)
+
+	if len(roleNames) == 0 {
+		logging.Logger.Debug("No IAM Service Linked Roles to nuke")
+		return nil
+	}
+
+	// NOTE: we don't need to do pagination here, because the pagination is handled by the caller to this function,
+	// based on IAMRoles.MaxBatchSize, however we add a guard here to warn users when the batching fails and has a
+	// chance of throttling AWS. Since we concurrently make one call for each identifier, we pick 100 for the limit here
+	// because many APIs in AWS have a limit of 100 requests per second.
+	if len(roleNames) > 100 {
+		logging.Logger.Debugf("Nuking too many IAM Service Linked Roles at once (100): halting to avoid hitting AWS API rate limiting")
+		return TooManyIamRoleErr{}
+	}
+
+	// There is no bulk delete IAM Roles API, so we delete the batch of IAM roles concurrently using go routines
+	logging.Logger.Debugf("Deleting all IAM Service Linked Roles in region %s", region)
+	wg := new(sync.WaitGroup)
+	wg.Add(len(roleNames))
+	errChans := make([]chan error, len(roleNames))
+	for i, roleName := range roleNames {
+		errChans[i] = make(chan error, 1)
+		go deleteIamServiceLinkedRoleAsync(wg, errChans[i], svc, roleName)
+	}
+	wg.Wait()
+
+	// Collect all the errors from the async delete calls into a single error struct.
+	var allErrs *multierror.Error
+	for _, errChan := range errChans {
+		if err := <-errChan; err != nil {
+			allErrs = multierror.Append(allErrs, err)
+			logging.Logger.Debugf("[Failed] %s", err)
+		}
+	}
+	finalErr := allErrs.ErrorOrNil()
+	if finalErr != nil {
+		return gruntworkerrors.WithStackTrace(finalErr)
+	}
+
+	for _, roleName := range roleNames {
```

Comment:
Out of curiosity - do we need this second loop through `roleNames` or do you think the log line in `deleteIamServiceLinkedRole` is sufficient (currently line 82)? They both say the same think, but this log line includes the `[OK]` prefix and region.

Reply:
Fixed in ec4c03c

Nah, it's duplicate. I've removed line 82 to make it consistent with what is going on in `iam_role.go`.